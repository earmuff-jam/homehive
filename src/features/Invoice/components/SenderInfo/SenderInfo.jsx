import React, { useEffect, useState } from "react";

import { useForm } from "react-hook-form";
import { useNavigate } from "react-router-dom";

import dayjs from "dayjs";

import { Stack } from "@mui/material";
import CustomSnackbar from "common/CustomSnackbar/CustomSnackbar";
import RowHeader from "common/RowHeader/RowHeader";
import {
  useGetSenderInfoQuery,
  useUpsertSenderInfoMutation,
} from "features/Api/invoiceApi";
import UserInfoViewer from "features/Invoice/components/UserInfo/UserInfoViewer";
import { useAppTitle } from "hooks/useAppTitle";

export default function SenderInfo() {
  useAppTitle("Sender Information");
  const navigate = useNavigate();
  const [showSnackbar, setShowSnackbar] = useState(false);

  const { data: senderInfo, isLoading: isSenderInfoLoading } =
    useGetSenderInfoQuery();

  const [
    upsertSenderInfo,
    {
      isLoading: isUpsertSendingInfoLoading,
      isSuccess: isUpsertSendingInfoSuccess,
    },
  ] = useUpsertSenderInfoMutation();

  const {
    register,
    handleSubmit,
    formState: { errors, isValid },
    reset,
  } = useForm({
    mode: "onChange",
    defaultValues: {
      first_name: "",
      last_name: "",
      email_address: "",
      phone_number: "",
      street_address: "",
      city: "",
      state: "",
      zipcode: "",
    },
  });

  const submit = (formData) => {
    formData["updatedOn"] = dayjs();
    upsertSenderInfo(formData);
  };

  useEffect(() => {
    if (isUpsertSendingInfoSuccess) {
      setShowSnackbar(true);
    }
  }, [isUpsertSendingInfoLoading]);

  useEffect(() => {
    if (senderInfo) {
      reset({
        first_name: senderInfo.first_name,
        last_name: senderInfo.last_name,
        email_address: senderInfo.email_address,
        phone_number: senderInfo.phone_number,
        street_address: senderInfo.street_address,
        city: senderInfo.city,
        state: senderInfo.state,
        zipcode: senderInfo.zipcode,
        updatedOn: senderInfo.updatedOn,
      });
    }
  }, [isSenderInfoLoading]);

  return (
    <Stack spacing={1} alignItems="center" data-tour={"sender-0"}>
      <RowHeader
        title="Add details about the sender"
        caption="Required fields are marked with an * "
      />
      <UserInfoViewer
        title="Sender Information"
        caption="Add details about the sender"
        register={register}
        errors={errors}
        isDisabled={!isValid}
        onSubmit={handleSubmit(submit)}
        loading={isUpsertSendingInfoLoading}
      />
      <CustomSnackbar
        showSnackbar={showSnackbar}
        setShowSnackbar={setShowSnackbar}
        title="Changes saved."
        caption="View Invoice"
        onClick={() => navigate("/view")}
      />
    </Stack>
  );
}
