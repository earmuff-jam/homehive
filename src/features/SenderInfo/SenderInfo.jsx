import { useEffect, useState } from "react";
import { BLANK_INDIVIDUAL_INFORMATION_DETAILS } from "../../common/UserInfo/constants";
import UserInfoViewer from "../../common/UserInfo/UserInfoViewer";
import RowHeader from "../../common/RowHeader/RowHeader";
import { Stack } from "@mui/material";
import dayjs from "dayjs";

export default function SenderInfo() {
  const [formData, setFormData] = useState(
    BLANK_INDIVIDUAL_INFORMATION_DETAILS
  );

  const handleChange = (ev) => {
    const { id, value } = ev.target;
    const updatedFormData = { ...formData };
    let errorMsg = "";
    for (const validator of updatedFormData[id].validators) {
      if (validator.validate(value)) {
        errorMsg = validator.message;
        break;
      }
    }
    updatedFormData[id] = {
      ...updatedFormData[id],
      value,
      errorMsg,
    };
    setFormData(updatedFormData);
  };

  const submit = (ev) => {
    ev.preventDefault();
    const draftData = Object.entries(formData).reduce(
      (acc, [key, valueObj]) => {
        acc[key] = valueObj.value;
        return acc;
      },
      {}
    );
    draftData["updated_on"] = dayjs();
    localStorage.setItem("senderInfo", JSON.stringify(draftData));
  };

  useEffect(() => {
    const localValues = localStorage.getItem("senderInfo");
    const parsedValues = JSON.parse(localValues);
    if (parsedValues) {
      const draftProfileDetails = { ...BLANK_INDIVIDUAL_INFORMATION_DETAILS };
      draftProfileDetails.first_name.value = parsedValues.first_name;
      draftProfileDetails.last_name.value = parsedValues.last_name;
      draftProfileDetails.email_address.value = parsedValues.email_address;
      draftProfileDetails.phone_number.value = parsedValues.phone_number;
      draftProfileDetails.street_address.value = parsedValues.street_address;
      draftProfileDetails.city.value = parsedValues.city;
      draftProfileDetails.state.value = parsedValues.state;
      draftProfileDetails.zipcode.value = parsedValues.zipcode;
      draftProfileDetails.updated_on = parsedValues.updated_on;
      setFormData(draftProfileDetails);
    }
  }, []);

  return (
    <Stack spacing={1} alignItems="center">
      <RowHeader
        title="Add details about the sender"
        caption="Required fields are marked with an * "
      />
      <UserInfoViewer
        title="Sender Information"
        caption="Add details about the sender"
        formData={formData}
        handleChange={handleChange}
        onSubmit={submit}
      />
    </Stack>
  );
}
