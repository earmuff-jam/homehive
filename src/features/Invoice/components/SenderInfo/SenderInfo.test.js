import React from "react";

import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";

import SenderInfo from "./SenderInfo";
import { configureStore } from "@reduxjs/toolkit";
import { render } from "@testing-library/react";

// Mock dependencies
jest.mock("hooks/useAppTitle", () => ({
  useAppTitle: jest.fn(),
}));

jest.mock("common/CustomSnackbar/CustomSnackbar", () => ({
  __esModule: true,
  default: ({ showSnackbar }) =>
    showSnackbar ? <div data-testid="snackbar">Changes saved.</div> : null,
}));

jest.mock("common/RowHeader/RowHeader", () => ({
  __esModule: true,
  default: ({ title, caption }) => (
    <div>
      <h1>{title}</h1>
      <p>{caption}</p>
    </div>
  ),
}));

jest.mock("features/Invoice/components/UserInfo/UserInfoViewer", () => ({
  __esModule: true,
  default: ({ onSubmit }) => (
    <form
      onSubmit={(e) => {
        e.preventDefault();
        onSubmit({ first_name: "John" });
      }}
    >
      <button type="submit">Submit</button>
    </form>
  ),
}));

jest.mock("features/Api/invoiceApi", () => ({
  useGetSenderInfoQuery: () => ({
    data: {},
    isLoading: false,
    isSuccess: true,
  }),
  useUpsertSenderInfoMutation: () => [
    jest.fn(),
    { isLoading: false, isSuccess: false },
  ],
}));

describe("SenderInfo component", () => {
  it("renders correctly and matches snapshot", () => {
    const store = configureStore({ reducer: () => ({}) });

    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter>
          <SenderInfo />
        </MemoryRouter>
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
