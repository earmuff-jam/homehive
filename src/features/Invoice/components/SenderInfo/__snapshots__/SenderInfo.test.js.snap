// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SenderInfo component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root css-aha7cr-MuiStack-root"
    data-tour="sender-0"
  >
    <div>
      <h1>
        Add details about the sender
      </h1>
      <p>
        Required fields are marked with an * 
      </p>
    </div>
    <form>
      <button
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
